### PR TITLE
[Admin] Reduced query size while creating/editing promotions

### LINKED ISSUE #4916: Administration of PromotionBundle with more than 20 thousand products killing server
I have 22.503 total products. I have 1249 taxons total.

When I try to create a new promotion (/administration/promotion/create) every product and taxon is loaded and this takes about 100 - 140 seconds!

It will be better to use AJAX calling by Select2.

### PR DESCRIPTION
**Pull request by NeverResponse**
| Q               | A
| --------------- | ---
| Bug fix?        | no |
| New feature?    | no |
| BC breaks?      | yes |
| Related tickets | fixes #4916 |
| License         | MIT |

The petri dish consisted of 350 products and 7 taxa.

**Before:** (using normal choice types)

image

**After:** (switched to autocompletes)

image

### CODE REVIEW

File: src/Sylius/Bundle/AdminBundle/Resources/views/Product/Tab/_taxonomy.html.twig [hunk 1]
```diff
@@ -1,8 +1,8 @@
 <div class="ui tab" data-tab="taxonomy">
     <h3 class="ui dividing header">{{ 'sylius.ui.taxonomy'|trans }}</h3>
     <div class="field">
-        {{ form_row(form.mainTaxon, {'remote_url': path('sylius_admin_ajax_taxon_index'), 'remote_criteria_type': 'contains', 'remote_criteria_name': 'name'}) }}
+        {{ form_row(form.mainTaxon) }}
```
Comment (pamil) [inline src/Sylius/Bundle/AdminBundle/Resources/views/Product/Tab/_taxonomy.html.twig]: :+1:

File: src/Sylius/Bundle/CoreBundle/Form/Extension/ProductTypeExtension.php [hunk 2]
```diff
@@ -47,6 +47,9 @@ public function buildForm(FormBuilderInterface $builder, array $options)
                 'resource' => 'sylius.taxon',
                 'choice_name' => 'name',
                 'choice_value' => 'code',
+                'remote_path' => 'sylius_admin_ajax_taxon_index',
```
Comment (pjedrzejewski) [inline src/Sylius/Bundle/CoreBundle/Form/Extension/ProductTypeExtension.php]: This couples the core to admin bundle, that's why it was moved to the template.

**Comment by pamil** [comment]
It also fixes https://github.com/Sylius/Sylius/issues/4916.

File: src/Sylius/Behat/Context/Ui/Admin/ManagingPromotionsContext.php [hunk 3]
```diff
@@ -148,7 +148,7 @@ public function iAddTheHasTaxonRuleConfiguredWith(...$taxons)
         $this->createPage->addRule('Has at least one from taxons');
 
         foreach ($taxons as $taxon) {
-            $this->createPage->selectRuleOption('Taxons', $taxon, true);
+            $this->createPage->selectAutoCompleteRuleOption('Taxons', $taxon, true);
```
Comment (pamil) [inline src/Sylius/Behat/Context/Ui/Admin/ManagingPromotionsContext.php]: `AutoComplete` -> `Autocomplete`?

File: src/Sylius/Behat/Page/Admin/Promotion/CreatePage.php [hunk 4]
```diff
@@ -247,4 +272,43 @@ private function getCollectionItems($collection)
...
+     */
+    private function selectAutoCompleteValue(NodeElement $autoComplete, $value)
+    {
+        Assert::isInstanceOf($this->getDriver(), Selenium2Driver::class);
+
+        $isVisibleScript = sprintf(
+            '$(document.evaluate("%s", document, null, XPathResult.FIRST_ORDERED_NODE_TYPE, null).singleNodeValue).dropdown("is visible")',
+            $autoComplete->getXpath()
+        );
+        $isAnyAsyncActionInProgressScript = sprintf(
```
Comment (pamil) [inline src/Sylius/Behat/Page/Admin/Promotion/CreatePage.php]: `sprintf` not needed though.

File: src/Sylius/Behat/Page/Admin/Promotion/CreatePage.php [hunk 5]
```diff
@@ -247,4 +272,43 @@ private function getCollectionItems($collection)
 
         return $items;
     }
+
+    /**
+     * @param NodeElement $autoComplete
+     * @param string $value
+     */
+    private function selectAutoCompleteValue(NodeElement $autoComplete, $value)
```
Comment (pamil) [inline src/Sylius/Behat/Page/Admin/Promotion/CreatePage.php]: Isn't it generic enough to extract it to a service? Or is it bound to promotions' create page?

**Comment by pjedrzejewski** [comment]
Thanks Jasiek, nice work! 👍

**Pull request merged**